crates/bevy_ecs/src/schedule/stage.rs

```diff
@@ -391,87 +347,112 @@ fn sort_systems(systems: &mut Vec<impl SystemContainer>) -> Result<(), Dependenc
-                    if !dependencies.contains(&system_index) {
-                        dependencies.push(system_index);
+            match labelled_systems.get(label) {
+                Some(dependants) => {
+                    for dependant in dependants.ones() {
+                        graph
+                            .entry(dependant)
+                            .or_insert_with(HashMap::default)
+                            .entry(system_index)
+                            .or_insert_with(HashSet::default)
+                            .insert(label.clone());
                     }
                 }
-                None => return Err(DependencyGraphError::LabelNotFound(label.clone())),
+                None => warn!(
+                    "System {} wants to be before unknown system label: {:?}",
+                    systems[system_index].name(),
+                    label
+                ),
             }
         }
     }
-    Ok(graph)
+    graph
 }
 
 /// Generates a topological order for the given graph.
 fn topological_order(
     systems: &[impl SystemContainer],
-    graph: &HashMap<usize, Vec<usize>>,
+    graph: &HashMap<usize, HashMap<usize, HashSet<BoxedSystemLabel>>>,
 ) -> Result<Vec<usize>, DependencyGraphError> {
     fn check_if_cycles_and_visit(
         node: &usize,
-        graph: &HashMap<usize, Vec<usize>>,
+        graph: &HashMap<usize, HashMap<usize, HashSet<BoxedSystemLabel>>>,
         sorted: &mut Vec<usize>,
         unvisited: &mut HashSet<usize>,
-        current: &mut HashSet<usize>,
+        current: &mut Vec<usize>,
     ) -> bool {
         if current.contains(node) {
             return true;
         } else if !unvisited.remove(node) {
             return false;
         }
-        current.insert(*node);
-        for dependency in graph.get(node).unwrap() {
+        current.push(*node);
+        for dependency in graph.get(node).unwrap().keys() {
             if check_if_cycles_and_visit(dependency, &graph, sorted, unvisited, current) {
                 return true;
             }
         }
         sorted.push(*node);
-        current.remove(node);
+        current.pop();
         false
     }
     let mut sorted = Vec::with_capacity(graph.len());
-    let mut current = HashSet::with_capacity_and_hasher(graph.len(), Default::default());
+    let mut current = Vec::with_capacity(graph.len());
     let mut unvisited = HashSet::with_capacity_and_hasher(graph.len(), Default::default());
     unvisited.extend(graph.keys().cloned());
     while let Some(node) = unvisited.iter().next().cloned() {
         if check_if_cycles_and_visit(&node, graph, &mut sorted, &mut unvisited, &mut current) {
-            return Err(DependencyGraphError::GraphCycles(
-                current
-                    .iter()
-                    .map(|index| systems[*index].display_name())
-                    .collect::<Vec<_>>(),
+            let mut cycle = Vec::new();
+            for index in 0..current.len() - 1 {
```

Comment:
Could you perhaps use `current.windows(2)` instead of direct indexing?

Reply:
That's literally the first time I see that API. I'll see if it fits, thanks for bringing that up.

Reply:
Done in 2dcd38fbb9545d7b8ace541c8f56d40856b4a056. Feels less clumsy now, but is also more complicated.